### PR TITLE
Init Changelog Sync Script

### LINKED ISSUE #5451: Write a script that parses all the CHANGELOG.md files exist in each package and updates the one on the root
### This is not a Bug Report, Feature Request, or related to Documentation

- [X] I have searched the existing issues

### Is there an existing issue for this?

- [X] I have searched the existing issues

### What's up?

Would be useful to automate the update of the main `CHANGELOG.md` file, based on the updates of each `CHANGELOG.md` in each package.
As suggested here: https://github.com/web3/web3.js/pull/5435#discussion_r972574314

### Examples/References

_No response_

### PR DESCRIPTION
**Pull request by Spacesai1or**
Supersedes #5769 and closes #5451

### What This PR Does

This PR introduces the script-pseudo-package `changelog` (found under `scripts/changelog`). It currently supports the following commands:

- `sync`: Takes the changes from the `## [Unreleased]` sections of each package's `CHANGELOG.md` and adds it to the root `CHANGELOG.md`

The following commands will add an entry to a specific package's `CHANGELOG.md` under the provided header:

- `added`
- `changed`
- `deprecated`
- `removed`
- `fixed`
- `security`

Any other command will display the help table;

```bash
Invalid command, please refer to below table for expected commands:
┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────┬─────────────────────────────────────┬────────────────────────────────────────────────────────────┬───────────────────────────────┐
│ (index) │     name     │                                             description                                             │              arguments              │                          example                           │        commandFunction        │
├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────┼─────────────────────────────────────┼────────────────────────────────────────────────────────────┼───────────────────────────────┤
│    0    │    'sync'    │ 'Checks CHANGELOG.md for each package in ./packages/ for entries not included in root CHANGELOG.md' │                 []                  │                           'sync'                           │  [Function: syncChangelogs]   │
│    1    │   'added'    │               "Adds changelogEntry to packageName's CHANGELOG.md under Added header"                │ [ 'packageName', 'changelogEntry' ] │   'yarn changelog added [packageName] [changelogEntry]'    │ [Function: addChangelogEntry] │
│    2    │  'changed'   │              "Adds changelogEntry to packageName's CHANGELOG.md under Changed header"               │ [ 'packageName', 'changelogEntry' ] │  'yarn changelog changed [packageName] [changelogEntry]'   │ [Function: addChangelogEntry] │
│    3    │ 'deprecated' │             "Adds changelogEntry to packageName's CHANGELOG.md under Deprecated header"             │ [ 'packageName', 'changelogEntry' ] │ 'yarn changelog deprecated [packageName] [changelogEntry]' │ [Function: addChangelogEntry] │
│    4    │  'removed'   │              "Adds changelogEntry to packageName's CHANGELOG.md under Removed header"               │ [ 'packageName', 'changelogEntry' ] │  'yarn changelog removed [packageName] [changelogEntry]'   │ [Function: addChangelogEntry] │
│    5    │   'fixed'    │               "Adds changelogEntry to packageName's CHANGELOG.md under Fixed header"                │ [ 'packageName', 'changelogEntry' ] │   'yarn changelog fixed [packageName] [changelogEntry]'    │ [Function: addChangelogEntry] │
│    6    │  'security'  │              "Adds changelogEntry to packageName's CHANGELOG.md under Security header"              │ [ 'packageName', 'changelogEntry' ] │  'yarn changelog security [packageName] [changelogEntry]'  │ [Function: addChangelogEntry] │
└─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────────────────────┴────────────────────────────────────────────────────────────┴───────────────────────────────┘

```

### Usage

To use the script, in the project root, run `yarn changelog <command> <arguments>` e.g.:

- `yarn changelog sync`
- `yarn changelog added web3 "I added this change (#1234)"`
- `yarn changelog removed web3 "I removed something (#1235)"`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1137,4 +1137,8 @@ should use 4.0.1-alpha.0 for testing.
 
 ## [Unreleased]
 
+### Added
+
+#### web3-core
+
```
Comment (Spacesai1or) [inline CHANGELOG.md:1143]: Ran `yarn changelog sync` to make this change as current `4.x` `CHANGELOG.md` is missing these headers

image

File: fixtures/contracts/Basic.sol [hunk 2]
```diff
@@ -28,23 +28,11 @@ contract Basic {
...
-	function getValues()
-		public
-		view
-		returns (
-			uint256,
-			string memory,
-			bool
-		)
-	{
+	function getValues() public view returns (uint256, string memory, bool) {
```
Comment (Spacesai1or) [inline fixtures/contracts/Basic.sol:31]: These file changes are a result of running `yarn format`

File: fixtures/contracts/Greeter.sol [hunk 3]
```diff
@@ -3,29 +3,29 @@
 pragma solidity ^0.8.13;
 
 contract Greeter {
-    uint256 counter;
```
Comment (Spacesai1or) [inline fixtures/contracts/Greeter.sol:6]: These file changes are a result of running `yarn format`

File: packages/web3-core/src/web3_subscription_manager.ts [hunk 4]
```diff
@@ -23,7 +23,8 @@ import { Web3RequestManager, Web3RequestManagerEvent } from './web3_request_mana
 import { Web3SubscriptionConstructor } from './web3_subscriptions';
 
 type ShouldUnsubscribeCondition = ({
-	id: sub,
+	id,
+	sub,
```
Comment (Spacesai1or) [inline packages/web3-core/src/web3_subscription_manager.ts:27]: This error was reported after running `yarn lint:fix`

File: packages/web3-eth-contract/test/fixtures/contracts/SampleStorageContract.sol [hunk 5]
```diff
@@ -18,17 +18,16 @@ along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
 pragma solidity ^0.8.7;
 
 contract SampleStorageContract {
+	uint256 uintNum;
```
Comment (Spacesai1or) [inline packages/web3-eth-contract/test/fixtures/contracts/SampleStorageContract.sol:21]: These file changes are a result of running `yarn format`

File: packages/web3-eth/src/web3_eth.ts [hunk 6]
```diff
@@ -1514,7 +1514,9 @@ export class Web3Eth extends Web3Context<Web3EthExecutionAPI, RegisteredSubscrip
 			subscription instanceof LogsSubscription &&
 			name === 'logs' &&
 			typeof args === 'object' &&
+			// @ts-expect-error Property 'fromBlock' does not exist on type '(ConstructorParameters<RegisteredSubscription[T]>[0] & object) | (ConstructorParameters<RegisteredSubscription[T]>[0] & null)'.
 			args.fromBlock &&
+			// @ts-expect-error Property 'fromBlock' does not exist on type '(ConstructorParameters<RegisteredSubscription[T]>[0] & object) | (ConstructorParameters<RegisteredSubscription[T]>[0] & null)'.
```
Comment (spacesailor24) [inline packages/web3-eth/src/web3_eth.ts:1519]: These errors were being reported when running `yarn lint:fix`

File: packages/web3-validator/src/utils.ts [hunk 7]
```diff
@@ -28,7 +28,7 @@ import { isAbiParameterSchema } from './validation/abi';
 import { isHexStrict } from './validation/string';
 // import { abiToJsonSchemaCases } from '../test/fixtures/abi_to_json_schema';
 
-export const parseBaseType = <T = typeof VALID_ETH_BASE_TYPES[number]>(
+export const parseBaseType = <T = (typeof VALID_ETH_BASE_TYPES)[number]>(
```
Comment (Spacesai1or) [inline packages/web3-validator/src/utils.ts:31]: These file changes are a result of running `yarn format`

**Review by jdevcs — APPROVED** [review]

**Review by avkos — APPROVED** [review]

**Review by nikoulai — APPROVED** [review]

**Comment by Muhammad-Altabba** [comment]
> 
> ### What This PR Does
> This PR introduces the script-pseudo-package `changelog` (found under `scripts/changelog`). It currently supports the following commands:
> 
> * `sync`: Takes the changes from the `## [Unreleased]` sections of each package's `CHANGELOG.md` and adds it to the root `CHANGELOG.md`
> 
> The following commands will add an entry to a specific package's `CHANGELOG.md` under the provided header:
> 
> * `added`
> * `changed`
> * `deprecated`
> * `removed`
> * `fixed`
> * `security`
> 
> Any other command will display the help table;
> 
> ```shell
> Invalid command, please refer to below table for expected commands:
> ┌─────────┬──────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────┬─────────────────────────────────────┬────────────────────────────────────────────────────────────┬───────────────────────────────┐
> │ (index) │     name     │                                             description                                             │              arguments              │                          example                           │        commandFunction        │
> ├─────────┼──────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────┼─────────────────────────────────────┼────────────────────────────────────────────────────────────┼───────────────────────────────┤
> │    0    │    'sync'    │ 'Checks CHANGELOG.md for each package in ./packages/ for entries not included in root CHANGELOG.md' │                 []                  │                           'sync'                           │  [Function: syncChangelogs]   │
> │    1    │   'added'    │               "Adds changelogEntry to packageName's CHANGELOG.md under Added header"                │ [ 'packageName', 'changelogEntry' ] │   'yarn changelog added [packageName] [changelogEntry]'    │ [Function: addChangelogEntry] │
> │    2    │  'changed'   │              "Adds changelogEntry to packageName's CHANGELOG.md under Changed header"               │ [ 'packageName', 'changelogEntry' ] │  'yarn changelog changed [packageName] [changelogEntry]'   │ [Function: addChangelogEntry] │
> │    3    │ 'deprecated' │             "Adds changelogEntry to packageName's CHANGELOG.md under Deprecated header"             │ [ 'packageName', 'changelogEntry' ] │ 'yarn changelog deprecated [packageName] [changelogEntry]' │ [Function: addChangelogEntry] │
> │    4    │  'removed'   │              "Adds changelogEntry to packageName's CHANGELOG.md under Removed header"               │ [ 'packageName', 'changelogEntry' ] │  'yarn changelog removed [packageName] [changelogEntry]'   │ [Function: addChangelogEntry] │
> │    5    │   'fixed'    │               "Adds changelogEntry to packageName's CHANGELOG.md under Fixed header"                │ [ 'packageName', 'changelogEntry' ] │   'yarn changelog fixed [packageName] [changelogEntry]'    │ [Function: addChangelogEntry] │
> │    6    │  'security'  │              "Adds changelogEntry to packageName's CHANGELOG.md under Security header"              │ [ 'packageName', 'changelogEntry' ] │  'yarn changelog security [packageName] [changelogEntry]'  │ [Function: addChangelogEntry] │
> └─────────┴──────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────────────────────┴────────────────────────────────────────────────────────────┴───────────────────────────────┘
> ```
> 
> ### Usage
> To use the script, in the project root, run `yarn changelog <command> <arguments>` e.g.:
> 
> * `yarn changelog sync`
> * `yarn changelog added web3 "I added this change (#1234)"`
> * `yarn changelog removed web3 "I removed something (#1235)"`

Great work @spacesailor24 ,
Could you please add this description to a `README.md` file inside `scripts/changelog`. And also add a summary of the usage to the main `README.md` file (with a link to the `README.md` file inside `scripts/changelog`)?
Thanks,

**Pull request merged**